### ISSUE TITLE
Bug with activation mode

### ISSUE THREAD

**Issue by jwashingtoncdm** [andbi/evrete#11]
In running some tests with evrete, a potential bug with the activation mode was uncovered. This bug is demonstrated in the following stand alone evrete program:

```
package com.tryit.evrete;

import org.evrete.KnowledgeService;
import org.evrete.api.ActivationMode;
import org.evrete.api.Knowledge;
import org.evrete.api.StatefulSession;

import java.util.Collection;
import java.util.LinkedList;
import java.util.concurrent.atomic.AtomicLong;

public class EVReteBug {


    public static class MyNumber {
        private int number;

        public MyNumber(int number) {
            this.number = number;
        }

        public int getNumber() {
            return number;
        }

        public void setNumber(int number) {
            this.number = number;
        }
    }

    public static void main(String[] args) {
        KnowledgeService service = new KnowledgeService();


        Knowledge knowledge = service.newKnowledge();

        knowledge.newRule("Add Odd Numbers")
                .forEach("$i", Integer.class)
                .where("($i % 2) != 0")
                .execute(ctx -> {
                    Integer i = ctx.get("$i");
                    ctx.insert(new MyNumber(i.intValue()));
                    System.out.println("Adding odd number "+i);
                });

        knowledge.newRule("Add Even Numbers")
                .forEach("$i", Integer.class)
                .where("($i % 2) == 0")
                .execute(ctx -> {
                    Integer i = ctx.get("$i");
                    ctx.insert(new MyNumber(i.intValue()));
                    System.out.println("Adding even number "+i);
                });

        knowledge.newRule("Sum My Numbers")
                .forEach("$n", MyNumber.class,"$sum", AtomicLong.class)
                .execute(ctx -> {
                    MyNumber n = ctx.get("$n");
                    AtomicLong sum = ctx.get("$sum");
                    sum.addAndGet(n.getNumber());
                    System.out.println("Summ "+sum.get());
                });



        Collection<Object> sessionData = new LinkedList<>();

        for(int i = 0; i < 10; i++) {
            sessionData.add(i);
        }
        AtomicLong sum = new AtomicLong(0);
        sessionData.add(sum);

        StatefulSession session = knowledge.newStatefulSession();
        //session.setActivationMode(ActivationMode.CONTINUOUS);
        session.insert(sessionData);
        session.fire();
        System.out.println("Sum " + sum.get());


        session.close();
        service.shutdown();
    }

}
```
The program sets up 3 rules, the first two rules inserts new facts (instances of MyNumber) depending on if a integer fact is even or odd. The last rule sums all the numbers into a AtomicLong. A stateful session is created and populated with 10 integers and a AtomicLong initialized to 0 and the rules are fired. With the activation set to  DEFAULT (by default), only the odd rules run. Setting the activation mode to CONTINUOUS causes both rules to fire. From my reading of the documentation, it appears this setting is something of optimization choice for the engine to decide whether to continue processing potentially "stale" or invalidated rules. However it appears this logic may be wrong as the above test case does not alter fact base in such a way to skip the processing the even rules. The new facts only affects the "downstream" rule for summing MyNumber facts. It seems that the decision to continue processing rules or not could be made by the engine based on if changes in the facts propagate changes to the rule currently being evaluated. In any case the activation mode setting should not cause the engine to produce erroneous results.

**Comment by andbi** [comment]
@jwashingtoncdm Hi,

Here is what is happening in the provided example:

1. Ten facts are inserted and buffered, waiting for the `fire()` command.
2. When you call `fire()`, the engine sees the buffered changes and determines which rules are affected. This stage is all about planning, and the planning is just about **fact types**, not their values.
3. When the activation plan is ready, the engine proceeds with rule activations. 
4. Activating the **"Add Odd Numbers"** rule introduces new changes in the working memory (five new facts), and the engine has to decide what to do next.
5. As you set activation mode to the `DEFAULT` mode, the initial activation plan is discarded, and the engine doesn't even get to the **"Add Even numbers"** rule. 

I totally understand why the described "planning" stage might seem redundant in this particular case, but it's crucial for more complex scenarios involving conditions across multiple fact types.

For example, if your condition were not just `($i % 2) == 0`, but rather `($i % 2) == 0 && $f > 0.0`, the situation wouldn't be that obvious, right? 

With such a condition, the rule would have to be fired not only when a new even number is inserted and a positive float already existed in the memory, but also for each new positive float and already existing even integers.

**Comment by jwashingtoncdm** [comment]
Hi, Thanks for your quick response and explanation, although I'm not convinced the engine is performing in a way that is expected or correct. I understand the Rete algorithm builds a directed graph to efficiently handle forward propagation of changes in facts and firing of dependent rules and needs to handle  complex interdependencies, such as the `($i % 2) == 0 && $f > 0.0` case you cite. However in this particular case, the insertion of new facts does not change any of the conditions for which both the odd and even rules should run. From a user perspective it was quite surprising (and counter intuitive) the "Add Even numbers" rule did not fire. In a complex system where this might've gone unnoticed, the results returned by the engine would be erroneous. While I'm sympathetic to the underlying complexities involved in the planning stage, etc., it's my opinion these are internal concerns of the engine and should never result in the engine producing a erroneous or misleading result.

**Comment by andbi** [comment]
> the insertion of new facts does not change any of the conditions for which both the odd and even rules should run

There's no way of knowing that in advance, like in the example I've shown. Or, if we decided to evaluate those conditions, it would defeat the whole idea of the Rate algorithm.

>should never result in the engine producing a erroneous or misleading result

Rule engines are always subjects to certain agreements. They're not procedural in a way that many developers are accustomed to. The `DEFAULT` mode, for that matter, is called "default" for a reason: if you author the very same ruleset in Drools or Jess (another two JSR-94 compliant rule engines), you will get exactly the same result. 

Regards,
@andbi

**Comment by andbi** [comment]
It seems like I've oversimplified my counter example. Consider this one instead:

`($i % 2) == 0 && $f > $i`

**Comment by jwashingtoncdm** [comment]
"if you author the very same ruleset in Drools or Jess (another two JSR-94 compliant rule engines), you will get exactly the same result." 

Well...actually, no.

Using this drl file:
```
import com.tryit.evrete.MyNumber
import java.util.concurrent.atomic.AtomicLong

rule "Add Odd Numbers"
	when 
		i: Integer( (intValue() % 2) != 0 )
	then
	    insert( new MyNumber(i.intValue()) );
	    System.out.println("Adding odd number "+i);
	end
rule "Add Even Numbers"
	when
		i: Integer( (intValue() % 2) == 0 )
	then
	    insert( new MyNumber(i.intValue()) );
	    System.out.println("Adding even number "+i);
	end
rule "Sum My Numbers"
	when 
		n: MyNumber()
		sum: AtomicLong()
	then
		sum.addAndGet(n.getNumber());
		System.out.println("Sum "+sum.get());
	end
```
And this driver code in Drools:
```

                KieServices ks = KieServices.Factory.get();
		KieContainer kcontainer = ks.getKieClasspathContainer();
		KieBase kbase = kcontainer.getKieBase();
		KieSession ksession = kbase.newKieSession();

		for(int i = 0; i < 10; i++) {
			ksession.insert(i);
		}
		AtomicLong sum = new AtomicLong(0);
		ksession.insert(sum);

		ksession.fireAllRules();

		System.out.println("Sum " + sum.get());
```
The rules work as expected. So I still maintain the default behavior in evrete is both unexpected and incorrect. Whether you agree with this assessment or not is entirely up to you, I merely brought up this issue in an effort to help improve this project in the spirit of open source and collaboration.

**Comment by andbi** [comment]
@jwashingtoncdm thanks for reporting the issue. Testing against Drools is an internal part of the build process, let me look into the modes used in those tests. They have long moved to the `evrete-benchmarks` module, but still git ignored for unknown reason. I'll open them too.

Btw, does the `CONTINUOUS` mode produce the same result as Drools?

**Comment by jwashingtoncdm** [comment]
Yes the `CONTINOUS`  mode seems to work in this example.

**Comment by andbi** [comment]
Got it. Somewhere along the way, the Drools tests have been split.
The `DEFAULT` mode has been tested against `delete()` and `update()` operations, and the `CONTINUOUS` one - against the `insert()` calls.

Many thanks for bringing up this issue. I confirm this inconsistency as a major bug, even though the library has never been intended to be a drop-in replacement for Drools and the spec leaves it up to the implementation to decide what is correct behavior.

As I see it, the only way out is to create a third mode that would be fully compatible with Drools. What do you think?

As for your case, you're safe as long as you stick with the `CONTINUOUS` mode and memory inserts. In terms of Drools compatibility, other memory operations are uncharted territory for now.

**Comment by jwashingtoncdm** [comment]
I only uncovered the mode setting feature in evrete when the "even number" rule unexpectedly didn't fire. Reading the documentation and the pseudo code around this feature led me to suspect that perhaps the engine was prematurely bailing when there were more rules to be processed. In my opinion the engine should support whatever the API allows, so if rules can insert/delete/update facts, then the engine should correctly process them. Users generally will not deep dive into the nuances of the engine configurations and various modes (unless they have to), so unless there is a good reason to expose this functionality, it's better to keep it hidden. This is especially true if a wrong setting causes the engine to misbehave, as in the case DEFAULT setting, since this kind of like a loaded gun the users can shoot themselves with. So if a "Drools" compatible mode covers the general case, then use that as the default setting and think about what, if any, settings should be user configurable.

**Comment by andbi** [comment]
Agree. The issue is fixed in 2.2.00, new documentation released https://www.evrete.org/docs/conflict-resolution/
